Add MaskGit support for iteratively predicting codebooks

rfejgin opened the pull request:
This PR adds support for MaskGit-like prediction of codebooks within a timestep (frame). It works as follows.

* After the main Magpie decoder we utilize second, smaller, Local Transformer (LT) to do the codebook-level predictions. This aspect similar to the previously integrated LT work and builds on that implementation.
* At training, some codebooks are randomly masked (i.e. replaced with the dedicated MASK token). A sequence consisting of Magpie's latent, the masked and the unmasked codes is given to the LT. The LT's task is to predict the correct value of the masked tokens.
* At inference, we:
  -  Start from a fully-masked token sequence
  -  While (any tokens are still masked):
      *  Predict the logits for all codebooks at once
      *  Sample from the logits
      *  Of the sampled codes pick the ones the model is most confident about (by logit value) and unmask those for the next round.

By the end of this algorithm all tokens (codebooks) have been unmasked / predicted. 

The two main potential benefits of this approach are:
- can be faster than fully-autoregressive LT prediction since every step of the loop above results in *multiple* codebooks being unmasked. It should also scale better to codecs with many codebooks.
- does not impose an arbitrate codebook generation order of 0..C-1, instead letting the model decide which codebooks it is most confident about.

By default, MaskGit training and sampling are both disabled. To enable it in training, set `use_maskgit: true` in the `model` section of the YAML config. To enable it at inference, add `--use_maskgit` to the `infer_and_evaluate.py` command line.

rfejgin committed:
Integrate Frechet Distance Metric in infer_and_evaluate.py

rfejgin committed:
Evaluate with frechet distance

rfejgin committed:
Merge remote-tracking branch 'blisc/magpietts_2503' into magpietts_2503_rfejgin

rfejgin committed:
MaskGit on codebook level (training)

 Implementation of MaskGit on codebook level - training only.

rfejgin committed:
MaskGit: make local transformer non-causal

rfejgin committed:
MaskGit sampling implementation

rfejgin committed:
MaskGit sampling working

rfejgin committed:
Fix top-k sampling bug

rfejgin committed:
Maskgit changes (WIP)

rfejgin committed:
Add some local manifest paths

rfejgin committed:
Merge remote-tracking branch 'blisc/magpietts_2503' into magpietts_2503_rfejgin

rfejgin committed:
Compatibility with embedding table layout changes

rfejgin committed:
Prepare MaskGit for PR

rfejgin committed:
Cleanup

rfejgin committed:
Merge remote-tracking branch 'blisc/magpietts_2503' into magpietts_2503_maskgit

rfejgin committed:
Fix typo in use of SpecialAudioTokens enum

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -250,9 +258,22 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
                 del state_dict[key]
         return state_dict
+    
+    def remap_names(self, state_dict):
```
Not directly related to this commit but this function has been useful to multiple folks - it helps with backwards compatibility to checkpoints from before code restructuring a few weeks ago. Hence merging it in (can remove it if necessary).

blisc replied:
Is this used in any of our scripts? I'm surprised older checkpoints worked without this change

rfejgin replied:
Older checkpoints (beyond ~a month ago) don't work without this; people have been making their own workaround, e.g. Shehzeen uses a snippet based on this to modify her checkpoints.

blisc replied:
This now collides with https://github.com/blisc/NeMo/pull/73 . Should we add this snippet to the model itself or keep it in infer and evaluate?

rfejgin replied:
I'm okay with keeping the version in #73; seems better to keep the main model file clean. I have removed `remap_names()` from this PR.

rfejgin commented on scripts/magpietts/evalset_config.py:
```diff
@@ -20,6 +20,21 @@
         'audio_dir' : '/datap/misc/Datasets/riva',
         'feature_dir' : '/datap/misc/Datasets/riva',
     },
+    'libri_dev_clean_eval_large': {
```
These are local paths on my machine but my understanding is that this file is where we put those local paths. Can remove if needed.

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -414,6 +499,105 @@ def logits_to_audio_codes(self, all_code_logits, audio_codes_lens):
 
         return all_preds
 
+    def maskgit_sample_codes(self, dec_output, temperature=0.7, topk=80, unfinished_items={}, finished_items={}, use_cfg=False, cfg_scale=1.0, n_steps=3):
```
MaskGit sampling

rfejgin committed:
PR preparation - cleanup and comments

rfejgin committed:
Remove frechet distance (not ready)

rfejgin committed:
Comments

rfejgin commented:
@blisc, @paarthneekhara could you review please? Thank you. And let me know if you'd like to get together to go over it.

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -63,6 +63,13 @@ def worker_init_fn(worker_id):
     dataset.text_conditioning_tokenizer = text_conditioning_tokenizer
 
 
+def cosine_schedule(x: torch.Tensor):
```
Can we make a new file under https://github.com/blisc/NeMo/tree/magpietts_2503/nemo/collections/tts/modules andmove this there?

rfejgin replied:
Sure, but what would you name such a file (would it be for magpie utilities or something more general)? Only for this small function? Options I was thinking of are either
- move it inside the `MagpieTTSModel` class as it's only used for MaskGit scheduling (but you could argue that other models besides Magpie might want to do that)
- put it in `nemo/collections/tts/parts/utils/helpers.py`? Or is that part of the structure that we hope to flatten away?

rfejgin replied:
@blisc for now I have moved this function inside the MagpieTTSModel class, is that okay with you? or if not could you suggest and answer to the questions above? thank you.

blisc replied:
I would either move it to parts or create another file under modules. I would not keep it inside the model.py file

rfejgin replied:
Moved to `magpietts_modules.py` in the modules directory

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -325,7 +346,7 @@ def get_speaker_embeddings(self, audio_16khz, audio_len_16khz):
             return speaker_embeddings
 
-    def compute_local_transformer_logits(self, dec_out, audio_codes_target):
+    def compute_local_transformer_logits(self, dec_out, audio_codes_target, targets_offset_by_one=False):
```
Can you docstring what targets_offset_by_one does?

rfejgin replied:
yes, done

rfejgin replied:
yes, done

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -343,7 +364,12 @@ def compute_local_transformer_logits(self, dec_out, audio_codes_target):
+            # for autoregressive local transformer the target for index 0 is codebook 0, for index 1 is codebook 1, etc.
+            local_transformer_output = local_transformer_output[:, :-1, :] # (B*T', C, E)
+        else:
+            # for MaskGit the target for index **1** is codebook 0, for index 2 is codebook 1, etc.
```
What's the target for index 0? Should it be codebook 0?
Shouldn't local_transformer_output be of size (x, C, E) not (x, C+1, E)?

rfejgin replied:
For MaskGit, index 0 does not participate in the loss. We only include input-masked tokens in the loss but cannot mask index zero since the input there is the Magpie latent. I made a diagram to try to clarify:

![lt_layout](https://github.com/user-attachments/assets/a506a438-c471-4f9e-a861-4cd486327a6c)

rfejgin replied:
For MaskGit index 0 is not included in the loss. We only include masked tokens (i.e. indices where the input to the transformer was replaced with a mask token) in the loss. But we cannot mask index 0 since that index is used to provide the Magpie latent, which we never want to mask. Unlike with autoregressive prediction, for MaskGit the inputs and the targets are always aligned, i.e. codebook 0 is at index 1 of the input of the transformer and also at index 1 of its output. For autoregressive, codebook 0 is at index 1 of the input of the transformer and at index 0 of the output. I made a diagram to try to clarify:
![lt_layout3](https://github.com/user-attachments/assets/274df176-9b1e-4445-97ab-5526d2182aab)

rfejgin replied:
Added this table in ascii format along with a dosctring in the latest commit

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -358,11 +384,67 @@ def compute_local_transformer_logits(self, dec_out, audio_codes_target):
+        codes_with_mask = torch.where(mask, self.mask_token_id, codes)
+        return codes_with_mask, mask
+
+    def compute_loss(self, logits, audio_codes, audio_codes_lens, mask_tokens_mask=None):
```
Can you docstring this function and explain when and how it is used?

rfejgin replied:
yes, done in latest commit

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -184,7 +192,7 @@ class SpecialAudioToken(Enum):
                 sa_n_heads=self.cfg.get('local_transformer_n_heads', 1),
                 kernel_size=1,
-                is_causal=True,
+                is_causal=cfg.get('use_maskgit', False),
```
Should we make this into a self parameter? Maybe something like self.second_stage_transformer is one of [None, "LT", "MG"].

rfejgin replied:
Yeah, makes sense. But I'll need to coordinate with @paarthneekhara to make sure I don't break existing checkpoints

rfejgin replied:
In latest commits we now have an enumeration that defines the `local_transformer_type`:
```python
    class LocalTransformerType(StrEnum):
        NO_LT = "none"
        AR = "autoregressive"
        MASKGIT = "maskgit"
```

This actually helped me catch a bug resulting from the former loosely-defined config strings. Thanks @blisc for suggesting to restructure this config.

rfejgin commented on scripts/magpietts/evaluate_generated_audio.py:
```diff
@@ -127,18 +127,24 @@ def evaluate(manifest_path, audio_dir, generated_audio_dir, language="en", sv_mo
-            gt_audio_text = transcribe_with_whisper(whisper_model, whisper_processor, gt_audio_filepath, language, device)
-            gt_audio_text = process_text(gt_audio_text)
+        except Exception as e:
+            print("Error during ASR: {}".format(e))
```
Added error handling so that the whole evaluation script doesn't crash if ASR failed on one sample

rfejgin committed:
Comments and cleanup

- Added docstrings
- Moved cosine_schedule() function inside the MagpieTTSModel class
  (its only current user)

rfejgin committed:
More comments and cleanup

paarthneekhara commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -358,11 +392,81 @@ def compute_local_transformer_logits(self, dec_out, audio_codes_target):
+        """
+        return torch.cos(x * (torch.pi / 2))
+
+    def maskgit_create_random_mask(self, codes):
```
Is the mask ever "all zeros" during training? It seems like we should have that case - My understanding is during inference, we start with an all masked input and then iteratively predict all the codebooks.

rfejgin replied:
full masking should happen more than 1/8th (for 8 codebooks) of the time since the schedule function is concave, i.e. it "prefers" to mask more than less. But yes, we could consider refinements here like hand-designing a schedule or enforcing at least one "all-masked" case per batch. I plan to refine this in future iterations but for now I think it's good enough since it's likely happening pretty frequently.

rfejgin committed:
Combine maskgit and local transformer configs

We now use `local_transformer_type` to control the type of local transformer.

At inference, providing '--use_local_transformer' will activate either Autoregressive or MaskGit sampling,
whichever the checkpoint was trained with.

rfejgin committed:
Remove deprecated `use_local_transformer` from config

We now use `local_transformer_type` to control the type of local transformer (for both AR and MaskGit).

rfejgin committed:
Update remaining config files to use `local_transformer_type`

rfejgin committed:
infer_and_evaluate updates for MaskGit

We use `--use_local_transformer` to activate either Autoregressive or MaskGit sampling,
whichever the checkpoint was trained with.

rfejgin commented:
Changing back to **draft** state since one of the recent changes seems to have broken MaskGit inference. Will restore once this is fixed.

rfejgin commented:
Okay, I found the issue. It looks like MaskGit checkpoints were trained with a causal local transformer whereas I'd meant for it to be noncausal. It was due to an incorrect config string being checked when deciding on causal/noncausal. This happened both at training and inference, so it worked (though definitely suboptimally - MaskGit should noncausal). Once I straightened out the the config options during the PR cleanup, suddenly inference mismatched with training, resulting in bad quality inference. 

I will retrain a model to validate the fix.

rfejgin committed:
Add a printout for debug

rfejgin commented:
After retraining with the corrected local transformer causality (i.e. making it non-causal as intended) the checkpoints are functional again. From a quick check the models also appear robust to wider range of settings of the number of MaskGit sampling steps.

rfejgin commented:
@blisc I think this is basically ready for merge. Minor thing that remains is https://github.com/blisc/NeMo/pull/69#discussion_r2072826611 - can you take a look at the question? thank you.

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -172,7 +172,10 @@ class SpecialAudioToken(Enum):
-        if cfg.get('use_local_transformer', False):
+
+        self.local_transformer_type = self.LocalTransformerType(cfg.get('local_transformer_type', 'none').lower())
+        print(f"Local transformer type: {self.local_transformer_type}")
```
Can you switch to logging or remove this statement?
`logging.info(f"Local transformer type: {self.local_transformer_type}")` or `logging.debug(f"Local transformer type: {self.local_transformer_type}")`

rfejgin replied:
Done

blisc commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -240,6 +243,11 @@ class SpecialAudioToken(Enum):
+    class LocalTransformerType(StrEnum):
+        NO_LT = "none"
+        AR = "autoregressive"
+        MASKGIT = "maskgit"
```
Let's move the Enums to a module file. Both this and class SpecialAudioToken(Enum):

rfejgin replied:
Done - they are now in the modules directory in a new file named `magpitetts_modules.py`.

blisc reviewed (commented):
4 more comments from me

rfejgin committed:
Address PR comments

* Move some functionality to a new file in the modules direcotry, `magpietts_modules.py`.
* Remove reduncant `remap_names` method in `magpietts.py` (already addressed in another PR).

rfejgin commented:
I believe all comments have been addressed

rfejgin committed:
Cleanup

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -92,32 +92,28 @@ def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+            debugpy.listen(('0.0.0.0', 5678))  # You can change the port if needed
+            print('Waiting for debugger to attach...')
+            debugpy.wait_for_client()  # This will block execution until the debugger attaches
+            print('Debugger is attached!')
```
^^ cleaned up in the next commit

rfejgin committed:
Restore a comment that was accidentally removed.

rfejgin commented on nemo/collections/tts/models/magpietts.py:
```diff
@@ -250,9 +241,8 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
-
+    
     def load_state_dict(self, state_dict, strict=True):
-        # Override to load all the keys except _speaker_verification_model and _codec_model
```
This comment was not meant to be removed -- restored in a subsequent commit.

blisc commented on scripts/magpietts/evaluate_generated_audio.py:
```diff
@@ -127,18 +127,24 @@ def evaluate(manifest_path, audio_dir, generated_audio_dir, language="en", sv_mo
-                pred_text = asr_model.transcribe([pred_audio_filepath])[0].text
+
+        try:
+            if language == "en":
```
Add a test
